### ISSUE TITLE
Multiplayer test components should not be listening to web requests fired from tests

### ISSUE THREAD

**Issue by peppy** [ppy/osu#14019]
See https://github.com/ppy/osu/blob/a5a0f12e199f4fc2b7fe171f0879b4974e8744f3/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerRoomManager.cs#L43-L128

and

https://github.com/ppy/osu/blob/26cc4af87c03fe513f4b5d31e27fb61db7245519/osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayer.cs#L140-L154

We shouldn't be writing tests like this because it's not how the game handles web requests anywhere, and makes the tests very confusing to follow. The requests should likely be fired via a call on `TestMultiplayerRoomManager` or similar, rather than from the test scene directly. Or not at all. The idea of the `DummyAPIAccess.HandleRequest` isn't for an individual component to *handle* the incoming requests, but for a test setup to provide mock responses for components to *receive* via the standard request flow.

**Comment by smoogipoo** [comment]
I highly disagree. If the process `request -> handlerequest -> response` is broken, then DummyAPIAccess is fundamentally flawed. I see no wrong doing here.

**Comment by smoogipoo** [comment]
The test room manager handling requests is only because it’s a sane place to do it.

**Comment by peppy** [comment]
Before I take this discussion further, you can see what I mean about it being inverse of how requests are usually handled, right? Nowhere in the game does a component bind to `IAPIAccess` and handle requests like is being done here. It's a new kind of invisible event flow which I haven't seen before and therefore very hard to follow.

**Comment by smoogipoo** [comment]
What's your suggestion, to move the `HandleRequest` binding to test scenes so test scenes individually handle the interconnectivity between the room manager and the multiplayer client? No thanks. It's a very specific flow which absolutely should not be in the test scenes, it's SPECIFICALLY crafted to match the server. Where it's placed is irrelevant - but that it's placed in one common location shared between all test scenes is absolutely relevant.

I don't disagree that the `API.Queue()` line _could_ be a direct call on the room manager, but I don't see any wrong with doing it as it is.

**Comment by smoogipoo** [comment]
Also I don't understand what you mean by "the game" when you refer to it multiple times. The "game" never handles requests in this way, that's why it's in TestRoomManager and not RoomManager. The "game" never binds to HandleRequest, but it _does_ do `API.Queue()` as that test step is doing.

**Comment by peppy** [comment]
There are two ways forward I can see.

- Fire the request from the `TestMultiplayerRoomManager` via an external call, and move the local operations to `.Success` event handlers (ie. addressing the "matching how we handle requests elsewhere" part). Specifically, lines like this:

https://github.com/ppy/osu/blob/a5a0f12e199f4fc2b7fe171f0879b4974e8744f3/osu.Game/Tests/Visual/Multiplayer/TestMultiplayerRoomManager.cs#L63

- Remove all of the API interactions here, as I don't believe they are required in the first place, nor does it seem to be testing anything specifically.

**Comment by smoogipoo** [comment]
I don't get it, sorry. Maybe you're best just coding something up because I can't follow what you're saying. That `Rooms.Add()` line you linked is the "server-side" room list. What API interactions are you talking about? the part where `GetBeatmapSetRequest` is requeued onto the game's API?

**Comment by peppy** [comment]
Right, that changes a lot. Maybe we need to rename that test component, or at least add a comment as to its intentions. Right now it's very unclear.

So in addition, I'd definitely propose we change these steps which are firing requests to instead being direct calls on the `TestMultiplayerRoomManager`:

https://github.com/ppy/osu/blob/32e29d042894e36ea86654dc855b8544a3d2721d/osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayer.cs#L141-L155

Something like `TestMultiplayerRoomManager.AddRoom()` should suffice fine?

**Comment by smoogipoo** [comment]
Sure, I'd be fine with an `AddRoom()`.
